src/wavelet_matrix.rs

```diff
@@ -0,0 +1,559 @@
+        if depth == self.bit_length {
+            return pos + rank;
+        }
+        let mut pos = pos;
+        let mut rank = rank;
+        let bit = Self::get_msb(val, depth, self.bit_length);
+        let rsbv = &self.layers[depth];
+        if bit {
+            let zeros = rsbv.num_zeros();
+            pos = rsbv.rank1(pos) + zeros;
+            rank = self.select_helper(rank, val, pos, depth + 1);
+            rsbv.select1(rank - zeros)
+        } else {
+            pos = rsbv.rank0(pos);
+            rank = self.select_helper(rank, val, pos, depth + 1);
+            rsbv.select0(rank)
+        }
+    }
+
+    /// Gets `k+1`-th smallest value in the given `range`.
+    ///
+    /// # Arguments
+    ///
+    /// - `range`: Position to be searched.
+    /// - `k`: Integer to be searched.
+    ///
+    /// # Complexity
+    ///
+    /// - $`O(\log \sigma)`$
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sucds::{WaveletMatrix, WaveletMatrixBuilder};
+    ///
+    /// let mut wmb = WaveletMatrixBuilder::new();
+    /// let s = "tobeornottobethatisthequestion";
+    /// s.chars().for_each(|c| wmb.push(c as usize));
+    ///
+    /// let wm = wmb.build().unwrap();
+    /// assert_eq!(wm.quantile(0..3, 0), 'b' as usize); // min in "tob" should be "b"
+    /// ```
+    pub fn quantile(&self, range: Range<usize>, k: usize) -> usize {
+        let mut val = 0;
+        let mut start_pos = range.start;
+        let mut end_pos = range.end;
+        let mut k = k;
+        for depth in 0..self.bit_length {
+            val <<= 1;
+            let rsbv = &self.layers[depth];
+            let zero_start_pos = rsbv.rank0(start_pos);
+            let zero_end_pos = rsbv.rank0(end_pos);
+            let zeros = zero_end_pos - zero_start_pos;
+            if k < zeros {
+                start_pos = zero_start_pos;
+                end_pos = zero_end_pos;
+            } else {
+                k -= zeros;
+                val |= 1;
+                start_pos = rsbv.num_zeros() + start_pos - zero_start_pos;
+                end_pos = rsbv.num_zeros() + end_pos - zero_end_pos;
+            }
+        }
+        val
+    }
+
+    /// Gets the all integers co-occurred more than `k` times in given `ranges`.
+    ///
+    /// # Arguments
+    ///
+    /// - `ranges`: A set of ranges to be searched.
+    /// - `k`: Occurrence threshold.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sucds::{WaveletMatrix, WaveletMatrixBuilder};
+    ///
+    /// let mut wmb = WaveletMatrixBuilder::new();
+    /// let s = "tobeornottobethatisthequestion";
+    /// s.chars().for_each(|c| wmb.push(c as usize));
+    ///
+    /// let wm = wmb.build().unwrap();
+    /// let ranges = vec![0..3, 3..6];
+    /// assert_eq!(wm.intersect(ranges, 2), vec!['o' as usize]);
+    /// ```
+    pub fn intersect(&self, ranges: Vec<Range<usize>>, k: usize) -> Vec<usize> {
```

Comment:
```suggestion
    /// assert_eq!(wm.intersect(&ranges, 2), vec!['o' as usize]);
    /// ```
    pub fn intersect(&self, ranges: &[Range<usize>], k: usize) -> Vec<usize> {
```

It would be better to get the argument as a slice because it doesn't force user to copy or move the vector.

Reply:
fixed in 390c840